pkg/products/threescale/reconciler.go

```diff
@@ -342,6 +396,343 @@ func (r *Reconciler) backupSystemSecrets(ctx context.Context, serverClient k8scl
+							}},
+						}},
+					},
+				},
+			},
+		},
+	}
+
+	// Setting GRPC
+	clusterName := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"cluster_name": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: "ratelimit",
+				},
+			},
+		},
+	}
+
+	grpcService := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"timeout": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: "2s",
+				},
+			},
+			"envoy_grpc": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: clusterName,
+				},
+			},
+		},
+	}
+	grpcInnerService := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"grpc_service": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: grpcService,
+				},
+			},
+		},
+	}
+	httpFilterGrpc := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			"domain": &structpb.Value{
+				Kind: &structpb.Value_StringValue{
+					StringValue: apicastRatelimiting,
+				},
+			},
+			"stage": &structpb.Value{
+				Kind: &structpb.Value_NumberValue{
+					NumberValue: 0,
+				},
+			},
+			"rate_limit_service": &structpb.Value{
+				Kind: &structpb.Value_StructValue{
+					StructValue: grpcInnerService,
+				},
+			},
+		},
+	}
+
+	// Setting up connection manager
+	manager := &hcm.HttpConnectionManager{
+		CodecType:  hcm.HttpConnectionManager_AUTO,
+		StatPrefix: "ingress_http",
+		RouteSpecifier: &hcm.HttpConnectionManager_RouteConfig{
+			RouteConfig: &envoyapi.RouteConfiguration{
+				Name:         "local_route",
+				VirtualHosts: []*v2route.VirtualHost{&virtualHost},
+			},
+		},
+		HttpFilters: []*hcm.HttpFilter{
+			{
+				Name:       "envoy.rate_limit",
+				ConfigType: &hcm.HttpFilter_Config{Config: httpFilterGrpc},
+			},
+			{
+				Name: "envoy.router",
+			},
+		},
+	}
+
+	pbst, err := ptypes.MarshalAny(manager)
+	if err != nil {
+		return fmt.Errorf("failed to convert HttpConnectionManager for rate limiting: %v", err)
+	}
+
+	listener := &envoyapi.Listener{
+		Name: "http",
+		Address: &envoycore.Address{
+			Address: &envoycore.Address_SocketAddress{
+				SocketAddress: &envoycore.SocketAddress{
+					Protocol: envoycore.SocketAddress_TCP,
+					Address:  "0.0.0.0",
+					PortSpecifier: &envoycore.SocketAddress_PortValue{
+						PortValue: uint32(8443),
+					},
+				},
+			},
+		},
+		FilterChains: []*listener.FilterChain{{
+			Filters: []*listener.Filter{{
+				Name:       "envoy.http_connection_manager",
+				ConfigType: &listener.Filter_TypedConfig{TypedConfig: pbst},
+			}},
+		}},
+	}
+
+	// Converting to Json and then to Yaml before creating the CR
+	rateLimitClusterJson, err := ResourcesToJSON(&rateLimitCluster)
+
+	yamlRateLimitCluster, err := yaml.JSONToYAML(rateLimitClusterJson)
+
+	clusterJson, err := ResourcesToJSON(&cluster)
+
+	yamlCluster, err := yaml.JSONToYAML(clusterJson)
+
+	listenerJson, err := ResourcesToJSON(listener)
+
+	yamlListener, err := yaml.JSONToYAML(listenerJson)
+
```

Comment:
Could we handle the above errors here also.

Reply:
Thanks @laurafitzgerald, updated the PR now